src/ert/config/observations.py

```diff
@@ -1,471 +1,243 @@
+
+    def __len__(self) -> int:
+        return len(self.gen_data_keys)
+
+    def to_xarray(self) -> xr.Dataset:
+        return (
+            pd.DataFrame(
+                data={
+                    "name": self.gen_data_keys,
+                    "obs_name": self.obs_names,
+                    "report_step": self.report_steps,
+                    "index": self.indexes,
+                    "observations": self.observations,
+                    "std": self.stds,
+                }
+            )
+            .set_index(["name", "obs_name", "report_step", "index"])
+            .to_xarray()
+        )
+
+
+class _GenObsAccumulator:
+    def __init__(self) -> None:
+        self.ds: _GenObsDataset = _GenObsDataset()
+
+    def write(
+        self,
+        gen_data_key: str,
+        obs_name: str,
+        report_step: int,
+        observations: List[float],
+        stds: List[float],
+        indexes: List[int],
+    ) -> None:
+        self.ds.gen_data_keys.extend([str(gen_data_key)] * len(observations))
+        self.ds.obs_names.extend([str(obs_name)] * len(observations))
+        self.ds.report_steps.extend([report_step] * len(observations))
+
+        self.ds.observations.extend(observations)
+        self.ds.stds.extend(stds)
+        self.ds.indexes.extend(indexes)
+
+
+class _SummaryObsAccumulator:
+    def __init__(self) -> None:
+        self.ds: _SummaryObsDataset = _SummaryObsDataset()
+
+    def write(
+        self,
+        summary_key: str,
+        obs_names: List[str],
+        observations: List[float],
+        stds: List[float],
+        times: List[Union[int, datetime]],
+    ) -> None:
+        self.ds.summary_keys.extend([str(summary_key)] * len(obs_names))
+        self.ds.obs_names.extend(map(str, obs_names))
+        self.ds.observations.extend(observations)
+        self.ds.stds.extend(stds)
+        self.ds.times.extend(times)
+
+
+# Columns used to form a key for observations of the response_type
+ObservationsIndices = {"summary": ["time"], "gen_data": ["index", "report_step"]}
+
+
 class EnkfObs:
     def __init__(self, obs_vectors: Dict[str, ObsVector], obs_time: List[datetime]):
         self.obs_vectors = obs_vectors
         self.obs_time = obs_time
-        self.datasets: Dict[str, xr.Dataset] = {
-            name: obs.to_dataset([]) for name, obs in sorted(self.obs_vectors.items())
+
+        vecs: List[ObsVector] = [*self.obs_vectors.values()]
+
+        gen_obs = _GenObsAccumulator()
+        sum_obs = _SummaryObsAccumulator()
+
+        # Faster to not create a single xr.Dataset per
+        # observation and then merge/concat
+        # this just accumulates 1d vecs before making a dataset
+        for vec in vecs:
+            if vec.observation_type == ResponseTypes.GEN_DATA:
+                for report_step, node in vec.observations.items():
+                    # wrt typing:
+                    # Goal is to deprecate the entire enkfobs, it has
```

Comment:
Is this an immediate goal? Otherwise I would reword this first part, as these comments have a tendency to come back and bite us 😅